### DISCUSSION TITLE
Welcome to Cars-Island-On-Azure Discussions!

### DISCUSSION THREAD

**Discussion by Daniel-Krzyczkowski** [Daniel-Krzyczkowski/Cars-Island-On-Azure#1] [General]
Do not hesitate to ask any questions you have about this project!

**Comment by sukesh13chukki** [comment]
Hello @Daniel-Krzyczkowski 

Greeting of the Day!

i am able to launch on localhost. But when I click on Login link I get an error(attached).
![Capture](https://user-images.githubusercontent.com/55776707/117528144-4df65a80-afee-11eb-9c45-5707a6882250.PNG)

Also, Please check my config files if I am missing something

![WebApp](https://user-images.githubusercontent.com/55776707/117528244-d4ab3780-afee-11eb-9449-1f971c4c6ddf.PNG)
![webAPI](https://user-images.githubusercontent.com/55776707/117528245-d5dc6480-afee-11eb-83eb-f2fece8445d1.PNG)

**Reply by Daniel-Krzyczkowski** [reply]
Hi @sukesh13chukki !
The reason why you see this error is the fact that you did not complete the configuration for the Azure AD B2C.
You have to provide all parameters under "AzureAdB2C" section in the appsettings.json file.
The error you see is generated by the identity library because it cannot retrieve proper information.
I hope this will help you.

**Reply by sukesh13chukki** [reply]
Thank you, let me do necessary changes. But i did not find ClientSecret on the azure portal.

**Reply by sukesh13chukki** [reply]
Hi @Daniel-Krzyczkowski 

Now that you started Azure identity series which is one of services in Cars-Island. So I would request you to create series of all services used in Cars Island. Which will give idea to implement that project with hands-on people like me.

**Reply by Daniel-Krzyczkowski** [reply]
Please check my blog once again, you will find posts about Cars Island implementation. :)

**Reply by sukesh13chukki** [reply]
Yes Sir,

i have started off with that series, but it is not in detail for example there is said to include client secret key but I don't know where to fetch it from . So I am stuck .

**Reply by Daniel-Krzyczkowski** [reply]
I understand, sorry to hear that some parts are not described in details. If you will need additional help, please let me know here. I try to provide as much details as possible in these blog articles.

**Reply by sukesh13chukki** [reply]
Hello @Daniel-Krzyczkowski 

I am in step 2 . where we configure B2C, registration, scope ,policy etc.

But when I tried to launch API, swagger is not loading. It is throwing 500 error . I know we haven't configured cosmos, storageaccount and all. So I commented out that in startup.cs as follow . Later I learnt that file appsettings.Development.json is not preset in specific path in src folder.  What is the purpose of this.


            services
                //.AddAppConfiguration(Configuration)
                    .AddAuthenticationWithAuthorizationSupport(Configuration)
                    //.AddDataServices()
                  //  .AddStorageServices()
                  //  .AddMessagingServices()
                    .AddSwagger();

**Reply by sukesh13chukki** [reply]
After commenting out   //  .AddJsonFile("appsettings.Development.json", optional: false, reloadOnChange: true)

it worked. But, now how to authenticate swagger, where can I generate JWT Token.

**Reply by Daniel-Krzyczkowski** [reply]
Thank you, this was my mistake. "appsettings.Development.json" file should be marked as optional as it is used only during development on local machine. "Optional" parameter should be set to true like presented below:

`.AddJsonFile("appsettings.Development.json", optional: true, reloadOnChange: true)`

When it comes to token generation. You will need to create Azure AD B2C tenant and register application in it. Here is the link to documentation. You can see step by step tutorial how to create Azure AD B2C tenant, setup Identity Experience Framework and how to test it:

[Create an Azure Active Directory B2C tenant](https://docs.microsoft.com/en-us/azure/active-directory-b2c/tutorial-create-tenant)

**Reply by sukesh13chukki** [reply]
No problem.  I will do necessary change . And, right now I am running it on local. Haven't pushed to app service.

I have created tenant and registered my application. Again, need to check on token.

**Reply by Daniel-Krzyczkowski** [reply]
Great to know. Do you use user flows or custom policies with your Azure AD B2C instance?

**Reply by sukesh13chukki** [reply]
I created a custom policy as you mentioned in second step.  i.e. signup_signin

**Reply by Daniel-Krzyczkowski** [reply]
Great. I assume that you registered web application (web portal) and web API in the Azure AD B2C (two apps). Did you grant permission to access API from the portal app (using "Expose API" tab and "API permissions" tab)?

**Reply by sukesh13chukki** [reply]
I yet to register web portal, Expose API you mean to say, creating a scope right. I have created it.

**Reply by Daniel-Krzyczkowski** [reply]
So now to quickly generate access token to access the API follow below steps. In you case you have to select web portal app in the first box and under "Access token" make sure you selected scope of your API. Then you can click "Run now". Once you register or sign in, you will get access token. Make sure you add https://jwt.ms as redirect URL for your portal app to get token directly in the browser URL. In the URL there will be ID Token returned together with Access token. Make sure you copy the right one because jwt.ms page will display ID Token by default (but you need the access token):

![gith1](https://user-images.githubusercontent.com/15244422/118759768-ea7fee80-b871-11eb-8abd-2ac65cd20704.PNG)

![gith2](https://user-images.githubusercontent.com/15244422/118759775-ece24880-b871-11eb-9b36-4e32953c31f6.PNG)

**Reply by sukesh13chukki** [reply]
Thanks for the elaborative steps . I have done couple things as you mentioned above. Let me run and grab token .

**Reply by sukesh13chukki** [reply]
Very sorry, I am using user flow , When I tried to Run UserFlow, I am not seeing any source to select

![image](https://user-images.githubusercontent.com/55776707/118791425-52efbf80-b8b4-11eb-9c39-634af2b0c982.png)


1. Web API

![image](https://user-images.githubusercontent.com/55776707/118791739-9fd39600-b8b4-11eb-85e8-43664f71ccd9.png)


2. Web Portal

![image](https://user-images.githubusercontent.com/55776707/118791878-c09beb80-b8b4-11eb-8839-fd38da72bbae.png)


In Web portal I haven't added scope and secret key, is it required .

**Reply by justoke** [reply]
Hi @sukesh13chukki 

I experienced almost identical results when trying to do this from the articles. You should be able to Run the user flow against the API app registration on it's own without the secret keys. If you are not redirected to the http://jwt.ms page with the token displayed, then something is not configured on the app registration(most likely the redirect urls). So you would check that the app registration has added the Web platform redirect URLs - if I recall from Daniel's article series, the actual interface changed since the article was published and I had trouble following and got things messed up.

Your app registration Authentication should look something like this and note I did not setup a client secret for the API:
![image](https://user-images.githubusercontent.com/26506785/118800322-adc6ee00-b897-11eb-9cf7-d163431aaafb.png)

Here is my api registration Run user flow view:
![image](https://user-images.githubusercontent.com/26506785/118799915-35602d00-b897-11eb-95c5-70e9ecfd5231.png)

I'd suggest starting with the API app registration.

**Reply by sukesh13chukki** [reply]
I Re-registered API, created secrete, added scope as well.

![image](https://user-images.githubusercontent.com/55776707/118804429-4e320800-b8c2-11eb-8c5d-ba39efdb6de2.png)

Also re-created UserFlow

but when i tried to run user-flow following is the thing, shows nothing on the list.

![image](https://user-images.githubusercontent.com/55776707/118804641-918c7680-b8c2-11eb-96f8-f0ab5faa5176.png)

**Reply by justoke** [reply]
Is there no application in the Application list? That is where all the app registrations should appear. For that to be the case, there would have to be no app registrations in the Azure AD B2C directory. Did you perhaps created the app registrations in the default Azure AD?

**Reply by sukesh13chukki** [reply]
Hello,

i have created under B2C tenant.

![image](https://user-images.githubusercontent.com/55776707/118950569-f1912480-b977-11eb-9f35-8610a9f033c7.png)

**Reply by sukesh13chukki** [reply]
@Daniel-Krzyczkowski 

let me stay put here. I will complete your Azure cloud identity series and come back to this.

Thank you.
